File: python/gl.py
import struct
from collections import namedtuple
import random
from obj import Obj

V2 = namedtuple('vertex2d', ['x', 'y'])
V3 = namedtuple('vertex3d', ['x', 'y', 'z'])

r = None

def char(c):
  return struct.pack('=c', c.encode('ascii'))

def hword(w):
  #short
  return struct.pack('=h', w)

def word(w):
  #long
  return struct.pack('=l', w)

def color(r, g, b):
  return bytes([b, g, r])

BLACK = color(0, 0, 0)
WHITE = color(255, 255, 255)

class Renderer(object):
  def __init__(self):
    self.light = V3(0,0,1)
    self.width = 0
    self.height = 0
    self.current_color = WHITE
    self.clear_color = BLACK
    self.vph = 0
    self.vpw = 0
    self.vpx = 0
    self.vpy = 0

  def createWindow(self, width, height):
    self.width = width
    self.height = height
    self.clear()

  def setViewPort(self, x, y, width, height):
    self.vph = height
    self.vpw = width
    self.vpx = x
    self.vpy = y

  def setCurrentColor(self, color):
    self.current_color = color

  def clear(self):
    self.framebuffer = [[self.clear_color for x in range(self.width)] for y in range(self.height)]
    self.zbuffer = [[-1e100 for x in range(self.width)] for y in range(self.height)]
  
  def setClearColor(self, color):
    self.clear_color = color
  
  def write(self, filename):
    padding = self.width % 4


    f = open(filename, 'bw')

    # header 14
    f.write(char('B'))
    f.write(char('M'))
    f.write(word(14+40+3*(self.width*self.height)))
    f.write(word(0))
    f.write(word(14+40))

    # infoHeader 40
    f.write(word(40))
    f.write(word(self.width))
    f.write(word(self.height))
    f.write(hword(1))
    f.write(hword(24))
    f.write(word(0))
    f.write(word(3*(self.width*self.height)))
    f.write(word(0))
    f.write(word(0))
    f.write(word(0))
    f.write(word(0))

    # bitmap
    for y in range(self.height):
      for x in range(self.width):
        f.write(self.framebuffer[y][x])
      for p in range(padding):
        f.write(struct.pack('=x'))


    f.close()
  
  def render(self):
    self.write('image.bmp')

  def point(self, x, y, color = None):
    self.framebuffer[y][x] = color or self.current_color

  def drawVertex(self, x, y):
    x01 = (x+1)/2
    y01 = (y+1)/2
    windowX = int(x01 * self.vpw + self.vpx)
    windowY = int(y01 * self.vph + self.vpy)
    self.point(windowX, windowY)
  
  def deNormalizedLine(self, x0, y0, x1, y1):
    dy = abs(y1 - y0)
    dx = abs(x1 - x0)
    steep = dy > dx

    if steep:
      x0, y0 = y0, x0
      x1, y1 = y1, x1 #swap en c++
      dy, dx = dx, dy


    offset = 0
    threshold =  dx
    y = y0

    for x in range(x0, x1, 1 if x1 > x0 else -1):

      if steep:
        self.point(y, x)
      else:
        self.point(x, y)

      offset += 2 * dy

      if offset >= threshold:
        if y0 < y1: y += 1
        else: y -= 1
        threshold += 2 * dx
  
  def vertexLine(self, A, B):
    x0 = A.x
    x1 = B.x
    y0 = A.y
    y1 = B.y
    self.deNormalizedLine(x0, y0, x1, y1)

  def line(self, nx0, ny0, nx1, ny1):
    x0 = int((nx0+1)/2 * self.vpw + self.vpx)
    x1 = int((nx1+1)/2 * self.vpw + self.vpx)
    y0 = int((ny0+1)/2 * self.vph + self.vpy)
    y1 = int((ny1+1)/2 * self.vph + self.vpy)
    self.deNormalizedLine(x0, x1, y0, y1)

  def triangleWireframe(self, A, B, C):
    self.vertexLine(A, B)
    self.vertexLine(B, C)
    self.vertexLine(C, A)
  
  def triangle(self, A, B, C, color=color(0, 0, 0), normalized=False):
    if normalized:
      A = V3(int((A.x + 1)/2 * self.vpw + self.vpx), int((A.y + 1)/2 * self.vph + self.vpy), A.z)
      B = V3(int((B.x + 1)/2 * self.vpw + self.vpx), int((B.y + 1)/2 * self.vph + self.vpy), B.z)
      C = V3(int((C.x + 1)/2 * self.vpw + self.vpx), int((C.y + 1)/2 * self.vph + self.vpy), C.z)
    xmin, xmax, ymin, ymax = self.bbox(A, B, C)

    for x in range(xmin, xmax + 1):
      for y in range(ymin, ymax + 1):
        P = V2(x, y)
        w, v, u = self.barycentric(A, B, C, P)
        if w < 0 or v < 0 or u < 0:
          continue

        z = A.z * w + B.z * v + C.z * u
        if z > self.zbuffer[y][x]:
          self.point(x, y, color=color)
          self.zbuffer[y][x] = z

  def bbox(self, A, B, C):
    xmin = A.x
    xmax = A.x
    if B.x < xmin: xmin = B.x
    if C.x < xmin: xmin = C.x

    if B.x > xmax: xmax = B.x
    if C.x > xmax: xmax = C.x

    ymin = A.y
    ymax = A.y
    if B.y < ymin: ymin = B.y
    if C.y < ymin: ymin = C.y

    if B.y > ymax: ymax = B.y
    if C.y > ymax: ymax = C.y
    return xmin, xmax, ymin, ymax

  def cross(self, v0, v1):
    cx = v0.y * v1.z - v0.z * v1.y
    cy = v0.z * v1.x - v0.x * v1.z
    cz = v0.x * v1.y - v0.y * v1.x

    if cz == 0:
      pass
    return V3(cx, cy, cz)

  def barycentric(self, A, B, C, P):
    c = self.cross(V3(B.x - A.x, C.x - A.x, A.x - P.x), V3(B.y - A.y, C.y - A.y, A.y - P.y))
    cx = c.x
    cy = c.y
    cz = c.z

    if cz == 0: return 0, 0, 0

    u = cx / cz
    v = cy / cz
    w = 1 - u - v

    return w, v, u

  def sub(self, v0, v1):
    return V3(
      v0.x - v1.x,
      v0.y - v1.y,
      v0.z - v1.z
    )

  def length(self, v0):
    return (v0.x**2 + v0.y**2 + v0.z**2) ** 0.5
  
  def norm(self, v0):
    l = self.length(v0)
    if l == 0: return V3(0, 0, 0)

    return V3(
      v0.x/l,
      v0.y/l,
      v0.z/l
    )
  
  def dot(self, v0, v1):
    return v0.x * v1.x + v0.y * v1.y + v0.z * v1.z

  def load(self, filename, translate, scale):
    model = Obj(filename)
    tx = translate[0]
    ty = translate[1]
    sx = scale[0]
    sy = scale[1]
    def calcTriangle(A, B, C):
      dA = V3((A.x + tx)*sx, (A.y + ty)*sy, A.z * 1e90)
      dB = V3((B.x + tx)*sx, (B.y + ty)*sy, B.z * 1e90)
      dC = V3((C.x + tx)*sx, (C.y + ty)*sy, C.z * 1e90)
      normal = self.norm(self.cross(
        self.sub(B, A),
        self.sub(C, A)
      ))

      intensity = self.dot(normal, self.light)

      grey = round(255 * intensity)

      if intensity >= 0:
        self.triangle(dA, dB, dC, color=color(grey, grey, grey), normalized=True)

    for face in model.faces:
      vertexNum = len(face)
      
      if vertexNum == 3:
        f1 = face[0][0] - 1
        f2 = face[1][0] - 1
        f3 = face[2][0] - 1
        A = V3(model.vertices[f1][0], model.vertices[f1][1], model.vertices[f1][2])
        B = V3(model.vertices[f2][0], model.vertices[f2][1], model.vertices[f2][2])
        C = V3(model.vertices[f3][0], model.vertices[f3][1], model.vertices[f3][2])
        
        calcTriangle(A, B, C)

      elif vertexNum == 4:
        f1 = face[0][0] - 1
        f2 = face[1][0] - 1
        f3 = face[2][0] - 1
        f4 = face[3][0] - 1
        A = V3(model.vertices[f1][0], model.vertices[f1][1], model.vertices[f1][2])
        B = V3(model.vertices[f2][0], model.vertices[f2][1], model.vertices[f2][2])
        C = V3(model.vertices[f3][0], model.vertices[f3][1], model.vertices[f3][2])
        D = V3(model.vertices[f4][0], model.vertices[f4][1], model.vertices[f4][2])
        

        calcTriangle(A, B, C)
        calcTriangle(A, C, D)

      else:
        for j in range(vertexNum):
          f1 = face[j][0]
          f2 = face[(j+1) % vertexNum][0]

          v1 = model.vertices[f1 - 1]
          v2 = model.vertices[f2 - 1]

          x1 = (v1[0] + translate[0]) * scale[0]
          y1 = (v1[1] + translate[1]) * scale[1]
          x2 = (v2[0] + translate[0]) * scale[0]
          y2 = (v2[1] + translate[1]) * scale[1]
          if (x1 >= -1 and y1 >= -1 and x2 >= -1 and y2 >= -1) and (x1 <= 1 and y1 <= 1 and x2 <= 1 and y2 <= 1):
            self.line(x1, y1, x2, y2)
  
  


#GLs start

def glInit():
  global r
  r = Renderer()

def glCreateWindow(width, height):
  if r:
    if width > 0 and height > 0:
      r.createWindow(width, height)
    else:
      print("Invalid values for window size")
  else:
    print("Missing initialization")

def glViewPort(x, y, width, height):
  if r and r.width > 0 and r.height > 0:
    if (x + width > r.width or y + height > r.height) and (x > 0 and y > 0 and width > 0 and height > 0):
      print("Cannot set viewport bigger than window")
    r.setViewPort(x, y, width, height)
  else:
    print("Bad window")

def glClear():
  if r and r.width > 0 and r.height > 0:
    r.clear()
  else:
    print("Bad window")

def glClearColor(R, G, B):
  if r and r.width > 0 and r.height > 0:
    if (R >= 0 and G >= 0 and B >= 0) and (R <= 1 and G <= 1 and B <= 1):
      r.setClearColor(color(int(R*255), int(G*255), int(B*255)))
    else:
      print("Invalid values for color")
  else:
    print("Bad window")

def glVertex(x, y):
  if r and r.width > 0 and r.height > 0:
    if (x >= -1 and y >= -1) and (x <= 1 and y <= 1):
      r.drawVertex(x, y)
    else:
      print("Cannot draw outside of viewport")
  else:
    print("Bad window")

def glColor(R, G, B):
  if r and r.width > 0 and r.height > 0:
    if (R >= 0 and G >= 0 and B >= 0) and (R <= 1 and G <= 1 and B <= 1):
      r.setCurrentColor(color(int(R*255), int(G*255), int(B*255)))
    else:
      print("Invalid values for color")
  else:
    print("Bad window")

def glLine(x0, y0, x1, y1):
  if r and r.width > 0 and r.height > 0:
    if (x0 >= -1 and y0 >= -1 and x1 >= -1 and y1 >= -1) and (x0 <= 1 and y0 <= 1 and x1 <= 1 and y1 <= 1):
      r.line(x0, y0, x1, y1)
    else:
      print("Invalid values for line limits")
  else:
    print("Bad window")

def glLoad(filename, translate, scale):
  if r and r.width > 0 and r.height > 0:
    r.load(filename, translate, scale)
  else:
    print("Bad window")

def glFinish():
  if r and r.width > 0 and r.height > 0:
    r.render()
  else:
    print("Bad window")





glInit()
glCreateWindow(1000, 1000)
glViewPort(50, 50, 900, 900)
glLoad("./models/model.obj", (0, 0), (0.75, 0.75))

glFinish()

File: cpp/gl.cpp
#include <iostream>
#include <fstream>
#include <stdlib.h>
#include <algorithm>
#include <vector>

#include "obj.h"

using namespace std;

class V2 {
	public:
	int x, y;
	V2(int x, int y){
		this->x = x;
		this->y = y;
	}
};

class V3 {
	public:
	int x, y, z;
	V3(int x, int y, int z){
		this->x = x;
		this->y = y;
		this->z = z;
	}
};

class Renderer {
	private:
	int width, height, vpx, vpy, vpw, vph;
	vector<vector<vector<unsigned char>>> framebuffer;
	unsigned char currentColor[3] = {255, 255, 255};
	unsigned char clearColor[3] = {0, 0, 0};

	void initVector(){
		for (int y = 0; y < height; y++){
			vector<vector<unsigned char>> line;
			for (int x = 0; x < width; x++){
				vector<unsigned char> color = {0, 0, 0};
				line.push_back(color);
			}
			framebuffer.push_back(line);
		}
	}
	
	void insertColor(int x, int y, unsigned char* col) {
		for (int i=0; i<3;i++){
			framebuffer[y][x][i] = col[2-i];
		}
	}
	
	void point(int x, int y){
		insertColor(x, y, currentColor);
	}

	void write(string filename){
		ofstream f;
		f.open(filename);
		//header
		static unsigned char header[14];
		
		int size = 14 + 40 + 3 * width * height;
		int offset = 14 + 40;
		int padding = width % 4;
		
		header[0] = (unsigned char)('B');
		header[1] = (unsigned char)('M');
		header[2] = (unsigned char)(size);
		header[3] = (unsigned char)(size >> 8);
		header[4] = (unsigned char)(size >> 16);
		header[5] = (unsigned char)(size >> 24);
		header[2] = (unsigned char)(0);
		header[3] = (unsigned char)(0);
		header[4] = (unsigned char)(0);
		header[5] = (unsigned char)(0);
		header[10] = (unsigned char)(offset);
		header[11] = (unsigned char)(offset >> 8);
		header[12] = (unsigned char)(offset >> 16);
		header[13] = (unsigned char)(offset >> 24);
		
		for (int i = 0; i < (int)(sizeof(header)/sizeof(header[0])); i++) {
			f << header[i];
		}
		
		//infoHeader
		static unsigned char infoHeader[40];
		int infoSize = 40;
		int bitmapSize = 3*width*height;
		
		infoHeader[0] = (unsigned char)(infoSize);
		infoHeader[1] = (unsigned char)(infoSize >> 8);
		infoHeader[2] = (unsigned char)(infoSize >> 16);
		infoHeader[3] = (unsigned char)(infoSize >> 24);
		infoHeader[4] = (unsigned char)(width);
		infoHeader[5] = (unsigned char)(width >> 8);
		infoHeader[6] = (unsigned char)(width >> 16);
		infoHeader[7] = (unsigned char)(width >> 24);
		infoHeader[8] = (unsigned char)(height);
		infoHeader[9] = (unsigned char)(height >> 8);
		infoHeader[10] = (unsigned char)(height >> 16);
		infoHeader[11] = (unsigned char)(height >> 24);
		infoHeader[12] = (unsigned char)(1);
		infoHeader[13] = (unsigned char)(0);
		infoHeader[14] = (unsigned char)(24);
		infoHeader[15] = (unsigned char)(0);
		infoHeader[16] = (unsigned char)(0);
		infoHeader[17] = (unsigned char)(0);
		infoHeader[18] = (unsigned char)(0);
		infoHeader[19] = (unsigned char)(0);
		infoHeader[20] = (unsigned char)(bitmapSize);
		infoHeader[21] = (unsigned char)(bitmapSize >> 8);
		infoHeader[22] = (unsigned char)(bitmapSize >> 16);
		infoHeader[23] = (unsigned char)(bitmapSize >> 24);
		infoHeader[24] = (unsigned char)(0);
		infoHeader[25] = (unsigned char)(0);
		infoHeader[26] = (unsigned char)(0);
		infoHeader[27] = (unsigned char)(0);
		infoHeader[28] = (unsigned char)(0);
		infoHeader[29] = (unsigned char)(0);
		infoHeader[30] = (unsigned char)(0);
		infoHeader[31] = (unsigned char)(0);
		infoHeader[32] = (unsigned char)(0);
		infoHeader[33] = (unsigned char)(0);
		infoHeader[34] = (unsigned char)(0);
		infoHeader[35] = (unsigned char)(0);
		infoHeader[36] = (unsigned char)(0);
		infoHeader[37] = (unsigned char)(0);
		infoHeader[38] = (unsigned char)(0);
		infoHeader[39] = (unsigned char)(0);
		
		for (int i = 0; i < (int)(sizeof(infoHeader)/sizeof(infoHeader[0])); i++) {
			f << infoHeader[i];
		}
		
		
		//bitmap
		for (int y = 0; y < (int)framebuffer.size(); y++) {
			for (int x = 0; x < (int)framebuffer[y].size(); x++) {
				for (int i=0; i<3;i++){
					f << framebuffer[y][x][i];
				}
			}
			for (int j = 0; j < padding; j++) f << (unsigned char) 0;
		}
		f.close();
		
	}
	
	void denormalizedLine(int x0, int y0, int x1, int y1){
		int dy = abs(y1 - y0);
		int dx = abs(x1 - x0);
		bool steep = dy > dx;
		
		if(steep){
			swap(x0, y0);
			swap(x1, y1);
			swap(dx, dy);
		}

		if(x0 > x1){
			swap(x0, x1);
			swap(y0, y1);
		}
		
		int offset = 0;
		int threshold = dx;
		int y = y0;		
		
		for(int x = x0; x <= x1; x++){
			if(steep){point(y, x);}
			else {point(x, y);}
			
			offset += 2 * dy;
			
			if(offset >= threshold){
				if(y0 < y1){y ++;}
				else {y--;}
				threshold += 2 * dx;
			}
		}
	}

	void vertexLine(V2 A, V2 B){
		this->denormalizedLine(A.x, A.y, B.x, B.y);
	}

	auto bbox(V2 A, V2 B, V2 C){
		int xmin = A.x;
		int xmax = A.x;
		if (B.x < xmin) xmin = B.x;
    if (C.x < xmin) xmin = C.x;

    if (B.x > xmax) xmax = B.x;
    if (C.x > xmax) xmax = C.x;

    int ymin = A.y;
    int ymax = A.y;
    if (B.y < ymin) ymin = B.y;
    if (C.y < ymin) ymin = C.y;

    if (B.y > ymax) ymax = B.y;
    if (C.y > ymax) ymax = C.y;

		struct retVals {int i1, i2, i3, i4;};
    return retVals{xmin, xmax, ymin, ymax};
	}

	auto cross(V3 v0, V3 v1){
		int cx = v0.y * v1.z - v0.z * v1.y;
		int cy = v0.z * v1.x - v0.x * v1.z;
    int cz = v0.x * v1.y - v0.y * v1.x;

		//NOTA
		struct retVals {int i1, i2, i3;};
		return retVals{cx, cy, cz};
	}

	auto barycentric(V2 A, V2 B, V2 C, V2 P){
		auto [cx, cy, cz] = cross(
			V3(B.x - A.x, C.x - A.x, A.x - P.x),
			V3(B.y - A.y, C.y - A.y, A.y - P.y)
		);
		float u = (float) cx / cz;
		float v = (float) cy / cz;
		float w = 1 - (u + v);
		struct retVals{float f1, f2, f3;};
		return retVals{w, v, u};
	}

	public:
	
	void createWindow(int width, int height){
		this->width=width;
		this->height=height;
		initVector();
	}
	
	void setViewPort(int x, int y, int width, int height){
		this->vpx = x;
		this->vpy = y;
		this->vpw = width;
		this->vph = height;
		
	}
	
	void setClearColor(unsigned char clearColor[3]){
		for (int i=0; i<3;i++){
			this->clearColor[i] = clearColor[i];
		}
	}
	
	void setCurrentColor(unsigned char color[3]){
		for (int i=0; i<3;i++){
			this->currentColor[i] = color[i];
		}
	}
	
	void clear(){
		for (int y = 0; y < (int)framebuffer.size(); y++) {
			for (int x = 0; x < (int)framebuffer[y].size(); x++) {
				insertColor(x, y, clearColor);
			}
		}
	}
	
	void drawVertex(float x, float y){
		int wx = (int)((x + 1)/2 * vpw + vpx);
		int wy = (int)((y + 1)/2 * vph + vpy);
		if(wx == vpw) wx -= 1;
		if(wy == vph) wy -= 1;
		point(wx, wy);
	}

	void drawLine(float nx0, float ny0, float nx1, float ny1){
		int x0 = (int)((nx0 + 1)/2 * vpw + vpx);
		int x1 = (int)((nx1 + 1)/2 * vpw + vpx);
		int y0 = (int)((ny0 + 1)/2 * vph + vpy);
		int y1 = (int)((ny1 + 1)/2 * vph + vpy);
		
		this->denormalizedLine(x0, y0, x1, y1);
		
	}
	
	void wireframeTriangle(V2 A, V2 B, V2 C){
		this->vertexLine(A, B);
		this->vertexLine(B, C);
		this->vertexLine(C, A);
	}

	void fillTriangle(V2 A, V2 B, V2 C){
		auto [xmin, xmax, ymin, ymax] = this->bbox(A, B, C);

		for(int x = xmin; x <= xmax; x++){
			for(int y = ymin; y <= ymax; y++){
				V2 P(x, y);
				auto [w, v, u] = this->barycentric(A, B, C, P);
				if(w < 0 | v < 0 | u < 0) continue;
				this->point(x, y);
			}
		}
	}

	void loadModel(string filename, float tx, float ty, float sx, float sy){
		Obj model(filename);
		
		for(int i = 0; i < (int)model.faces.size(); i++){
			vector<vector<int>> face = model.faces[i];
			
			for(int j = 0; j < (int)face.size(); j++){
				int f1 = face[j][0];
				int f2 = face[(j+1) % (int)face.size()][0];
				
				vector<float> v1 = model.vertices[f1 - 1];
				vector<float> v2 = model.vertices[f2 - 1];
				
				float x1 = (v1[0] + tx) * sx;
				float y1 = (v1[1] + ty) * sy;
				float x2 = (v2[0] + tx) * sx;
				float y2 = (v2[1] + ty) * sy;
				if((x1 >= -1 && y1 >= -1 && x2 >= -1 && y2 >= -1) && (x1 <= 1 && y1 <= 1 && x2 <= 1 && y2 <= 1)){
					drawLine(x1, y1, x2, y2);
				}
			}
		}
	}
	
	void render(){
		write("image.bmp");
	}
	
};

Renderer ren;

void glInit(){}

void glCreateWindow(int width, int height){
	ren.createWindow(width, height);
}

void glViewPort(int x, int y, int width, int height){
	ren.setViewPort(x, y, width, height);
}

void glClear() {ren.clear();}

void glClearColor(float r, float g, float b) {
	unsigned char color[] = {(unsigned char)(r*255), (unsigned char)(g*255), (unsigned char)(b*255)};
	ren.setClearColor(color);
}

void glVertex(float x, float y){
	ren.drawVertex(x, y);
}

void glColor(float r, float g, float b) {
	unsigned char color[] = {(unsigned char)(r*255), (unsigned char)(g*255), (unsigned char)(b*255)};
	ren.setCurrentColor(color);
}

void glLine(float x0, float y0, float x1, float y1){
	ren.drawLine(x0, y0, x1, y1);
}

void glLoad(string filename, float tx, float ty, float sx, float sy){
	ren.loadModel(filename, tx, ty, sx, sy);
}

void glFinish() {ren.render();}

int main() {
	glCreateWindow(1003, 1003);
	glViewPort(0, 0, 1000, 1000);
	glColor(0.5, 0, 0);

	//glLoad("./models/indoor plant_02.obj", 0, -3, 0.25, 0.25);

	ren.wireframeTriangle(V2(10, 70),  V2(50, 160), V2(70, 80));
	ren.wireframeTriangle(V2(180, 50), V2(150, 1),  V2(70, 180));
	ren.wireframeTriangle(V2(180, 150), V2(120, 160), V2(130, 180));

	glColor(0.5, 0, 0.5);
	ren.fillTriangle(V2(10, 70),  V2(50, 160), V2(70, 80));
	//ren.fillTriangle(V2(180, 50), V2(150, 1),  V2(70, 180));
	//ren.fillTriangle(V2(180, 150), V2(120, 160), V2(130, 180));
	//auto [u, v, w] = ren.barycentric(V2(10, 70),  V2(50, 160), V2(70, 80), V2(65, 100));
	//cout<< u << v << w;
	glFinish();
	
	return 0;
}
